### PR TITLE
Add encryption key derivation from PIN

### PR DESCRIPTION
**Pull request by sosthene-nitrokey**
<del>The test currently fail because of a bug in `cbor-smol` fixed in https://github.com/nickray/cbor-smol/pulls</del>

~The tests now fails because of the littlefs2 lookahead bug fixed in version 0.4~

### CODE REVIEW

**Comment by sosthene-nitrokey** [comment]
`SetPin` has a flag to enable a random key to be associated with the PIN. This key can be accessed after validation of the pin with `GetPinKey`. Changing the Pin with `ChangePin` doesn't change the key that is derived.

When creating a SoftwareAuth instance, the runner can give it some key material that would come from the hardware itself. This makes it slightly harder to extract data required for any brute-forcing.

Here is a pseudocode overview of the key derivation process.
```rust
// length depends on hardware
let device_ikm: [u8]= values_from_hardware();

// generated on first power-up, stays constant for the lifetime of the device
let device_salt: [u8;32] = csprng();

// The salt is useful for the security proof of HKDF
let device_extracted = hkdf_extract(salt: device_salt, ikm: device_ikm);

// There is domain separation between each app
for app_id in apps {
    let app_encryption_key: [u8; 32] = device_extracted.expand(info: app_id, 32);

    // Deriving keys from a PIN or Password provided by the user
    for pwd in passwords {
        let salt = csprng();
        // Per-pin key is fully random, then wrapped using the Pin
        let wrapped_key = csprng();

        // This is fine because app_encryption_key is uniform, therefore pin_key is too
        //
        // We can't use PBKDF or argon2 here because of limited hardware. 
        // Ideally such a step would be done on the host
        //
        // `pin_key` is never stored, and derived on each call to `CheckPin` and `GetPinKey`
        let pin_key = HMAC(key: app_encryption_key, pin_id || len(pwd) || pwd || salt);

        // On pin creation or change, the key is wrapped and stored on a persistent filesystem
        // The constant nonce is acceptable and won't lead to nonce reuse because the `pin_key` is only used to encrypt data once
        // Any change of pin changes the `pin_key` also changes the salt
        // which means that it is not possible to use the same key twice
        store(aead_encrypt(key: pin_key, data: wrapped_key, nonce: [0;12]))


        // On GetPinKey requests, this gets us the key
        aead_decrypt(key: pin_key, data: load(), nonce: [0;12]))
   } 
}
```

**Comment by robin-nitrokey** [comment]
Please add a patch for littlefs2 to fix the CI.  https://github.com/Nitrokey/littlefs2/releases/tag/v0.3.2-nitrokey-2 should work.

**Comment by sosthene-nitrokey** [comment]
This patch prevents trussed from compiling because of the `read(&mut self)` change

**Comment by sosthene-nitrokey** [comment]
It appears to work by using your littlefs 0.4 branch.

File: Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,11 @@ rand_core = "0.6.4"
...
-trussed = { git = "https://github.com/trussed-dev/trussed", rev = "1c55b3b2dd6a9e1cfc55758635baf0d0bbf387d1", features = ["serde-extensions"] }
+trussed = { version = "0.1.0", features = ["serde-extensions"] }
 
 [dev-dependencies]
 rand_core = { version = "0.6.4", default-features = false, features = ["getrandom"] }
-trussed = { git = "https://github.com/trussed-dev/trussed", rev = "1c55b3b2dd6a9e1cfc55758635baf0d0bbf387d1", features = ["serde-extensions", "virt"] }
+trussed = { version = "0.1.0", features = ["serde-extensions", "virt"] }
+
+[patch.crates-io]
+trussed = { git = "https://github.com/robin-nitrokey/trussed.git", rev = "f943d88aa43c72bc76c5c8bf8c8b5bb3638a4b85" }
```
Comment (robin-nitrokey) [inline Cargo.toml:28]: Use https://github.com/Nitrokey/trussed/releases/tag/v0.1.0-nitrokey-5 instead for easier readability and for reproducibility?

Comment (sosthene-nitrokey) [reply Cargo.toml]: Actually the branch we were currently using was missing https://github.com/trussed-dev/trussed/pull/94 so I fixed it. I thought it was the other way around

**Comment by sosthene-nitrokey** [comment]
After a bit of discussion with @szszszsz here is a clearer version of the pseudocode:

```rust
// length depends on hardware
let device_ikm: [u8]= values_from_hardware();

// generated on first power-up, stays constant for the lifetime of the device
let device_salt: [u8;32] = csprng();

// The salt is useful for the security proof of HKDF
let device_prk = hkdf_extract(salt: device_salt, ikm: device_ikm);

fn get_app_key(app_id) {
    // Domain separation between the apps
    // This means the `app_key` can also be used for purposes other than Pin key derivation.
    return hkdf_expand(prk: device_prk, info: app_id, output_len: 32);
}

fn register_pin(app_id, pin_id, pin) {
    let app_key = get_app_key(app_id);
    let salt = csprng();
    let key_to_be_wrapped = csprng();

    // Get a pseudo-random key from the pin and the salt
    //
    // This is fine because app_key is uniform, therefore pin_key is too
    // because HMAC is a PRF
    //
    // We can't use PBKDF or argon2 here because of limited hardware. 
    // Ideally such a step would be done on the host
    //
    // `pin_kek` is never stored
    let pin_kek = HMAC(key: app_key, pin_id || len(pin) || pin || salt);

    // On pin creation or change, the key is wrapped and stored on a persistent filesystem
    // The constant nonce is acceptable and won't lead to nonce reuse because the `pin_kek` is only used to encrypt this data once
    // 
    // Any change of pin changes also changes the salt
    // which means that it is not possible to get the `pin_kek` twice
    let wrapped_key = aead_encrypt(key: pin_kek, data: key_to_be_wrapped, nonce: [0;12]);

    to_presistent_storage(salt, wrapped_key);
}

fn get_pin_key(app_id, pin_id, pin) {
    let app_key = get_app_key(app_id);
    let (salt, wrapped_key) = from_persistent_storage();

    // re-derive the pin kek
    let pin_kek = HMAC(key: app_key, pin_id || len(pin) || pin || salt);

    // Unwrap the key
    let unwrapped_key = aead_decrypt(key: pin_kek, data: wrapped_key , nonce: [0;12]))
    return unwrapped_key;
}


fn change_pin(app_id, pin_id, old_pin, new_pin) {
    let app_key = get_app_key(app_id);
    let key_to_be_wrapped = get_pin_key(app_id, pin_id, pin);

    // The procedure is the same as for `register_pin` but it reuses the `key_to_be_wrapped` instead of generating it

    // Generate a new salt for the new pin
    let salt = csprng();

    let pin_kek = HMAC(key: app_key, pin_id || len(new_pin) || new_pin || salt);

    let wrapped_key = aead_encrypt(key: pin_kek, data: key_to_be_wrapped, nonce: [0;12]);

    to_presistent_storage(salt, wrapped_key);
}
```

The algorithm used are SHA-256 (for HKDF and HMAC) and ChaCha8Poly1305 (for AEAD). The CSPRNG comes from Trussed, so it's a `ChaCha8Rng` from the [ChaCha20](https://docs.rs/chacha20/0.8.2/chacha20/struct.ChaCha8Rng.html) crate

References:

- [HKDF](https://www.rfc-editor.org/rfc/rfc5869)
- [HMAC](https://www.rfc-editor.org/rfc/rfc2104)
- [ChaCha8Poly1305](https://eprint.iacr.org/2019/1492)

**Comment by szszszsz** [comment]
The reference to [ChaCha8Poly1305](https://eprint.iacr.org/2019/1492) seems to be wrong. Can you check that?
I would like to have it describing the key generation process.

**Comment by szszszsz** [comment]
Supporting diagram to the pseudocode:
image

**Comment by sosthene-nitrokey** [comment]
There isn't really a standard to point to for ChaCha8 so it points instead to a publication justifying reduced rounds. I also added the ChaCha20 RFC for completeness

**Comment by szszszsz** [comment]
In other words, response to my [request](https://github.com/trussed-dev/trussed-auth/pull/10#issuecomment-1441893276) is at:
- https://datatracker.ietf.org/doc/html/rfc8439#section-4
    > The Poly1305 key MUST be unpredictable to an attacker.  Randomly
   generating the key would fulfill this requirement, except that
   Poly1305 is often used in communications protocols, so the receiver
   should know the key.  Pseudorandom number generation such as by
   encrypting a counter is acceptable.  Using ChaCha with a secret key
   and a nonce is also acceptable.
- https://datatracker.ietf.org/doc/html/rfc8439#section-2.6 
    > As said in [Section 2.5](https://datatracker.ietf.org/doc/html/rfc8439#section-2.5), it is acceptable to generate the one-time
   Poly1305 key pseudorandomly.  This section defines such a method.
- https://datatracker.ietf.org/doc/html/rfc8439#section-2.5

**Comment by sosthene-nitrokey** [comment]
>     * https://datatracker.ietf.org/doc/html/rfc8439#section-4
>       > The Poly1305 key MUST be unpredictable to an attacker.  Randomly
>       > generating the key would fulfill this requirement, except that
>       > Poly1305 is often used in communications protocols, so the receiver
>       > should know the key.  Pseudorandom number generation such as by
>       > encrypting a counter is acceptable.  Using ChaCha with a secret key
>       > and a nonce is also acceptable.
> 
>     * https://datatracker.ietf.org/doc/html/rfc8439#section-2.6
>       > As said in [Section 2.5](https://datatracker.ietf.org/doc/html/rfc8439#section-2.5), it is acceptable to generate the one-time
>       > Poly1305 key pseudorandomly.  This section defines such a method.

These are irrelevant as we don't use Poly1305 directly but instead use it as part of the ChaCha20Poly1305 AEAD construction, so the Poly1305 key is generated using ChaCha: https://datatracker.ietf.org/doc/html/rfc8439#section-2.6 using the key and the nonce, making it unpredictable.

The relevant part applies to the nonce:

https://datatracker.ietf.org/doc/html/rfc8439#section-2.8

> A 96-bit nonce -- different for each invocation with the same key

In our case it is ok because the `pin_kek` is used for encryption only once on registration, so there is no risk of reuse.

File: src/lib.rs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Copyright (C) Nitrokey GmbH
 // SPDX-License-Identifier: Apache-2.0 or MIT
 
-#![no_std]
+// #![no_std]
```
Comment (robin-nitrokey) [inline src/lib.rs:4]: Can you please re-enable this?

Comment (sosthene-nitrokey) [reply src/lib.rs]: Done.

**Review by robin-nitrokey — CHANGES_REQUESTED** [review]
Looks good to me, though I did not  double-check the implementation details yet.

The only issue I see is that it is easy to make a mistake when choosing between the `SetPin` and `ChangePin` requests. Should we drop the `SetPin` request to avoid accidentally dropping the derived key?  See also: https://github.com/Nitrokey/trussed-secrets-app/pull/34#discussion_r1123125805

**Comment by robin-nitrokey** [comment]
I think we also talked about having a way to access the application key (or deriving a key from it) when no PIN is set.  That’s not implemented yet, right?

**Comment by sosthene-nitrokey** [comment]
> The only issue I see is that it is easy to make a mistake when choosing between the `SetPin` and `ChangePin` requests. Should we drop the `SetPin` request to avoid accidentally dropping the derived key? See also: [Nitrokey/trussed-secrets-app#34 (comment)](https://github.com/Nitrokey/trussed-secrets-app/pull/34#discussion_r1123125805)

We could instead make it fail when the Pin type is not a simple hash to ensure such bugs are obvious in testing. I still think we want to have it available.

> I think we also talked about having a way to access the application key (or deriving a key from it) when no PIN is set. That’s not implemented yet, right?

No that is not implemented. I think it can be a separate PR. Do we need it now?

**Comment by robin-nitrokey** [comment]
> We could instead make it fail when the Pin type is not a simple hash to ensure such bugs are obvious in testing.

Sounds good.

**Comment by sosthene-nitrokey** [comment]
Another solution I see would be to rename `SetPin` to `ResetPin`, so that the intent is clearer. This will also make more sense with policies when they are implemented.

**Comment by robin-nitrokey** [comment]
> Another solution I see would be to separate `SetPin` to `CreatePin` and `ResetPin`

Would there be a difference between the two requests?  Or just two names for the same request?

**Comment by sosthene-nitrokey** [comment]
`CreatePin` would fail if the Pin already exists. `ResetPin` would fail if it doesn't.

**Comment by robin-nitrokey** [comment]
> `CreatePin` would fail if the Pin already exists. `ResetPin` would fail if it doesn't.

Makes sense.  Do you want to add it to this PR or should we do it separately?

**Comment by sosthene-nitrokey** [comment]
We can do it in another PR.

**Review by robin-nitrokey — APPROVED** [review]

**Pull request merged**